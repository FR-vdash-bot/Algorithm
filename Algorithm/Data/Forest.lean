/-
Copyright (c) 2024 Yuyang Zhao. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Yuyang Zhao
-/
import Mathlib.Data.Set.Insert

inductive Forest (α : Type*)
  | nil : Forest α
  | node (a : α) (child sibling : Forest α) : Forest α
deriving Repr

namespace Forest
variable {α : Type*}

@[simp]
def roots : (f : Forest α) → Set α
  | .nil => ∅
  | .node a _ s => insert a s.roots

@[simp]
def support : (f : Forest α) → Set α
  | .nil => ∅
  | .node a c s => insert a (c.support ∪ s.support)

lemma roots_subset_support (f : Forest α) : f.roots ⊆ f.support := by
  induction f with
  | nil => rfl
  | node a c s _ ihs =>
    exact Set.insert_subset_insert (Set.subset_union_of_subset_right ihs _)

@[simp]
def pre : (f : Forest α) → List α
  | .nil => ∅
  | .node a c s => a :: c.pre ++ s.pre

@[simp]
def post : (f : Forest α) → List α
  | .nil => ∅
  | .node a c s => c.post ++ a :: s.post

@[simp]
protected def append (f g : Forest α) : Forest α :=
  match f with
  | .nil  => g
  | .node a c s => .node a c (s.append g)

instance : Append (Forest α) where
  append := Forest.append

end Forest
