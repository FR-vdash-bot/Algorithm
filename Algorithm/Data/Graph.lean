
import Algorithm.Data.Classes.AssocArray

structure Graph where
  V : Type _
  [instDecidableEqV : DecidableEq V]
  E : Type _
  fst' : E → V
  snd' : E → V
  StarList : Type _
  [instAssocArrayStarList : AssocArray V (List E) StarList]
  star' : StarList
  costar' : StarList
  fst_star' : ∀ v, ∀ e ∈ AssocArray.get star' v, fst' e = v
  snd_costar' : ∀ v, ∀ e ∈ AssocArray.get costar' v, snd' e = v

namespace Graph
variable {g : Graph}

instance : DecidableEq g.V := g.instDecidableEqV
instance : AssocArray g.V (List g.E) g.StarList := g.instAssocArrayStarList
-- instance : GetElem g.StarList g.V (List g.E) (fun _ _ ↦ True) := AssocArray.instGetElem
-- by infer_instance

def E.fst (e : g.E) : g.V := g.fst' e
def E.snd (e : g.E) : g.V := g.snd' e

def V.star (v : g.V) : List g.E := AssocArray.get g.star' v
def V.costar (v : g.V) : List g.E := AssocArray.get g.costar' v

lemma fst_star (v : g.V) : ∀ e ∈ v.star, e.fst = v := g.fst_star' v
lemma snd_costar (v : g.V) : ∀ e ∈ v.costar, e.snd = v := g.snd_costar' v

end Graph
