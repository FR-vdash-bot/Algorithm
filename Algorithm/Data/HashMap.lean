/-
Copyright (c) 2024 Yuyang Zhao. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Yuyang Zhao
-/
import Algorithm.Data.Classes.Dict

namespace Std.HashMap
variable {ι α : Type*} [BEq ι] [Hashable ι]

instance : LawfulEmptyCollection (HashMap ι α) ι where
  not_mem_empty _ := not_mem_empty

variable [LawfulBEq ι]

instance : ToFinset (HashMap ι α) ι where
  isEmpty := isEmpty
  isEmpty_iff_forall_not_mem := isEmpty_iff_forall_not_mem
  toFinset c := ⟨c.keys, by simpa [List.Nodup] using c.distinct_keys⟩
  mem_toFinset := by simp

instance : Size (HashMap ι α) where
  size := size
  size_eq_sizeTM := by simp [sizeTM_eq_card_toFinset, toFinset]

instance : Dict (HashMap ι α) ι α where
  setElem := insert
  valid_setElem := by simp
  getElem_setElem_self := by simp
  getElem_setElem_of_ne _ _ _ _ hij := by simp [getElem_insert, hij]
  erase := erase
  valid_erase := by simp
  getElem_erase_of_ne := by simp

end Std.HashMap
