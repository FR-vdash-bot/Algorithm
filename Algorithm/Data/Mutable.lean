/-
Copyright (c) 2024 Yuyang Zhao. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Yuyang Zhao
-/

/-!
# Mutable

-/

universe u v

-- waiting lean4#2292
/-
structure Mutable (α : Type u) : Type u where
  mk ::
  get : α

attribute [extern "lean_mk_Mutable"] Mutable.mk
attribute [extern "lean_Mutable_get"] Mutable.get

variable {α : Type u}

@[extern "lean_Mutable_get_with"]
protected def Mutable.getWith (x : @& Mutable α) (f : α → α) : α :=
  x.get
-/

structure Mutable (α : Type u) : Type u where
  private __mk__ ::
  private __get__ : α

namespace Mutable
variable {α : Type u} {β : Type v}

@[extern "lean_mk_Mutable"]
def mk (a : α) : Mutable α := __mk__ a

@[match_pattern, extern "lean_Mutable_get"]
def get (a : @& Mutable α) : α := __get__ a

set_option linter.unusedVariables false in
@[extern "lean_Mutable_modify"]
unsafe def modify (x : @& Mutable α) (f : α → α) : α :=
  x.get

set_option linter.unusedVariables false in
unsafe def getWithImpl (x : Mutable α)
    (f : α → α) (hf : ∀ a, f a = a) : α :=
  Mutable.modify x f

@[implemented_by Mutable.getWithImpl]
def getWith (x : Mutable α)
    (f : α → α) (hf : ∀ a, f a = a) : α :=
  f x.get

set_option linter.unusedVariables false in
@[extern "lean_Mutable_modify2"]
unsafe def modify₂ (x : @& Mutable α) (f : α → β) (g : β → α) : β :=
  f x.get

set_option linter.unusedVariables false in
unsafe def getWith₂Impl (x : Mutable α)
    (f : α → β) (g : β → α) (hgf : ∀ a, g (f a) = a) : β :=
  Mutable.modify₂ x f g

@[implemented_by Mutable.getWith₂Impl]
def getWith₂ (x : Mutable α)
    (f : α → β) (g : β → α) (hgf : ∀ a, g (f a) = a) : β :=
  f x.get

end Mutable
