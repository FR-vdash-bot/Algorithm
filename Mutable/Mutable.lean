/-
Copyright (c) 2024 Yuyang Zhao. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Yuyang Zhao
-/

/-!
# Mutable

-/

universe u v

-- waiting lean4#2292
/-
structure Mutable (α : Type u) : Type u where
  mk ::
  get : α

attribute [extern "lean_mk_Mutable"] Mutable.mk
attribute [extern "lean_Mutable_get"] Mutable.get

variable {α : Type u}

@[extern "lean_Mutable_get_with"]
protected def Mutable.getWith (x : @& Mutable α) (f : α → α) : α :=
  x.get
-/

structure Mutable (α : Type u) : Type u where
  private __mk__ ::
  private __get__ : α

namespace Mutable
variable {α : Type u} {β : Type v}

@[extern "lean_mk_Mutable"]
def mk (a : α) : Mutable α := __mk__ a

@[extern "lean_Mutable_get"]
def get (a : @& Mutable α) : α := __get__ a

theorem ext {x y : Mutable α} (get : x.get = y.get) : x = y :=
  match x, y, get with | ⟨_⟩, ⟨_⟩, rfl => rfl

theorem ext_iff {x y : Mutable α} : x = y ↔ x.get = y.get :=
  ⟨congrArg get, ext⟩

@[simp]
theorem mk_eq_mk {x y : α} : mk x = mk y ↔ x = y :=
  ext_iff

set_option linter.unusedVariables false in
@[extern "lean_Mutable_modify"]
unsafe def getModifyUnsafe (x : @& Mutable α) (f : α → α) : α :=
  x.get

set_option linter.unusedVariables false in
unsafe abbrev getModifyImpl (x : Mutable α)
    (f : α → α) (hf : ∀ a, f a = a) : α :=
  Mutable.getModifyUnsafe x f

@[implemented_by Mutable.getModifyImpl]
def getModify (x : Mutable α)
    (f : α → α) (hf : ∀ a, f a = a) : α :=
  f x.get

set_option linter.unusedVariables false in
@[extern "lean_Mutable_modify2"]
unsafe def getModify₂Unsafe (x : @& Mutable α) (f : α → β × α) : β :=
  (f x.get).fst

set_option linter.unusedVariables false in
unsafe abbrev getModify₂Impl (x : Mutable α)
    (f : α → β × α) (hgf : ∀ a, (f a).snd = a) : β :=
  Mutable.getModify₂Unsafe x f

@[implemented_by Mutable.getModify₂Impl]
def getModify₂ (x : Mutable α)
    (f : α → β × α) (hgf : ∀ a, (f a).snd = a) : β :=
  (f x.get).fst

end Mutable
