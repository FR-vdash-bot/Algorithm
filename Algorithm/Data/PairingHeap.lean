/-
Copyright (c) 2024 Yuyang Zhao. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Yuyang Zhao
-/
import Algorithm.Data.Classes.MinHeap
import Batteries.Data.PairingHeap

namespace Batteries

variable {α β : Type*} {le : α → α → Bool}

namespace PairingHeapImp

@[simp]
lemma Heap.size_nil :
    nil (α := α).size = 0 :=
  rfl

@[simp]
lemma Heap.size_node (a : α) (c s : Heap α) :
    (node a c s).size = c.size + 1 + s.size :=
  rfl

@[simp]
lemma Heap.foldTree_nil (nil : β) (join : α → β → β → β) :
    Heap.nil.foldTree nil join = nil :=
  rfl

@[simp]
lemma Heap.foldTree_node (nil : β) (join : α → β → β → β) (a : α) (c s : Heap α) :
    (node a c s).foldTree nil join = join a (c.foldTree nil join) (s.foldTree nil join) :=
  rfl

@[simp]
lemma Heap.toListUnordered_nil :
    nil.toListUnordered = ([] : List α) :=
  rfl

@[simp]
lemma Heap.toListUnordered_node (a : α) (c s : Heap α) :
    (node a c s).toListUnordered = a :: (c.toListUnordered ++ s.toListUnordered) := by
  unfold toListUnordered
  generalize hl : [] = l; nth_rw 2 [← hl]; clear hl
  induction c generalizing s l with
  | nil => rfl
  | node a c' s' ihc' ihs' =>
    simp? at * says simp only [foldTree_node, List.cons.injEq, true_and, List.cons_append] at *
    rw [ihc', ihc', ihs', List.append_assoc]

@[simp]
lemma Heap.length_toListUnordered (x : Heap α) :
    x.toListUnordered.length = x.size := by
  induction x with
  | nil => rfl
  | node _ _ _ _ _ => simp [*, Nat.add_right_comm]

attribute [simp] Heap.noSibling_merge

@[simp]
lemma Heap.merge_nil_right (x : Heap α) (hx : x.NoSibling) :
    x.merge le nil = x :=
  match x, hx with | nil, _ | node _ _ nil, _ => rfl

@[simp]
lemma Heap.coe_toListUnordered_merge_node_right
    (x : Heap α) (hx : x.NoSibling) (a : α) (c s : Heap α) :
    ((x.merge le (node a c s)).toListUnordered : Multiset α) =
      x.toListUnordered + (a ::ₘ c.toListUnordered : Multiset α) := by
  match x, hx with
  | nil, _ => rfl
  | node _ _ nil, _ =>
    simp_rw [merge]
    split_ifs <;>
      (simp_rw [toListUnordered_node, toListUnordered_nil, List.append_nil,
        ← Multiset.cons_coe, ← Multiset.coe_add, ← Multiset.singleton_add]; ac_rfl)

@[simp]
lemma Heap.coe_toListUnordered_merge_node (a₁ : α) (c₁ s₁ : Heap α) (a₂ : α) (c₂ s₂ : Heap α) :
    (((node a₁ c₁ s₁).merge le (node a₂ c₂ s₂)).toListUnordered : Multiset α) =
      a₁ ::ₘ a₂ ::ₘ c₁.toListUnordered + c₂.toListUnordered := by
  simp_rw [merge]
  split_ifs <;>
    (simp_rw [toListUnordered_node, toListUnordered_nil, List.append_nil,
      ← Multiset.cons_coe, ← Multiset.coe_add, ← Multiset.singleton_add]; ac_rfl)

lemma Heap.coe_toListUnordered_merge (x : Heap α) (y : Heap α)
    (hx : x.NoSibling) (hy : y.NoSibling) :
    ((x.merge le y).toListUnordered : Multiset α) = x.toListUnordered + y.toListUnordered := by
  match x, y, hx, hy with
  | nil, nil, _, _ => rfl
  | nil, node a c nil, _, _ =>
    rw [merge, toListUnordered_node, toListUnordered_nil, Multiset.coe_nil, zero_add]
  | node a c nil, nil, _, _ =>
    rw [merge, toListUnordered_node, toListUnordered_nil, Multiset.coe_nil, add_zero]
  | node a₁ c₁ nil, node a₂ c₂ nil, _, _ =>
    simp_rw [coe_toListUnordered_merge_node, toListUnordered_node, toListUnordered_nil,
      List.append_nil, ← Multiset.cons_coe, ← Multiset.singleton_add]; ac_rfl

theorem Heap.WF.noSibling {x : Heap α} (hx : x.WF le) : x.NoSibling :=
  match x, hx with
  | .nil, _ => .nil
  | .node _ _ .nil, _ => .node _ _

lemma Heap.le_of_nodeWF [t : IsTrans α (le · ·)] [TotalBLE le]
    (a : α) (x : Heap α) (hx : x.NodeWF le a) :
    ∀ b ∈ x.toListUnordered, le a b := by
  intro b hb
  induction x generalizing a with
  | nil => nomatch hb
  | node xa xc xs hc hs =>
    simp? at hb says simp only [toListUnordered_node, List.mem_cons, List.mem_append] at hb
    obtain ⟨hx, hxc, hxs⟩ := hx
    obtain (rfl | hb | hb) := hb
    · exact hx
    · have := hc xa hxc hb
      exact t.trans _ _ _ hx this
    · exact hs a hxs hb

lemma Heap.le_of_wf [r : IsRefl α (le · ·)] [t : IsTrans α (le · ·)] [TotalBLE le]
    (a : α) (c s : Heap α) (hx : (node a c s).WF le) :
    ∀ b ∈ (node a c s).toListUnordered, le a b := by
  intro b hb
  simp? at hb says simp only [toListUnordered_node, List.mem_cons, List.mem_append] at hb
  cases hx with
  | node hx =>
    obtain (rfl | hb | hb) := hb
    · exact r.refl b
    · exact Heap.le_of_nodeWF a c hx b hb
    · nomatch hb

@[simp]
lemma Heap.coe_toListUnordered_combine (x : Heap α) :
    ((x.combine le).toListUnordered : Multiset α) = x.toListUnordered := by
  induction x using Heap.combine.induct le with
  | case1 a₁ c₁ a₂ c₂ s ih =>
    match hc : combine le s, noSibling_combine le s with
    | .nil, _ =>
      simp? [combine, hc, - Multiset.coe_eq_coe, - Multiset.cons_coe, - Multiset.coe_add] says
        simp only [combine, hc, noSibling_merge, merge_nil_right,
          coe_toListUnordered_merge_node, Multiset.cons_add, toListUnordered_node]
      simp only [← Multiset.cons_coe, ← Multiset.coe_add, ← Multiset.singleton_add, ← ih, hc]
      ac_rfl
    | .node sa sc .nil, _ =>
      simp? [combine, hc, toListUnordered_node, - Multiset.coe_eq_coe, - Multiset.cons_coe,
          - Multiset.coe_add] says
        simp only [combine, hc, noSibling_merge,
          coe_toListUnordered_merge_node_right, coe_toListUnordered_merge_node, Multiset.cons_add,
          Multiset.add_cons, toListUnordered_node]
      simp only [← Multiset.cons_coe, ← Multiset.coe_add, ← Multiset.singleton_add, ← ih, hc,
        toListUnordered_node]
      ac_rfl
  | case2 x hx =>
    simp [combine]

end PairingHeapImp

namespace PairingHeap

@[simp]
lemma coe_toListUnordered_merge (x : PairingHeap α le) (y : PairingHeap α le) :
    ((x.merge y).toListUnordered : Multiset α) = x.toListUnordered + y.toListUnordered :=
  x.val.coe_toListUnordered_merge y.val x.prop.noSibling y.prop.noSibling

@[simp]
lemma coe_toListUnordered_singleton (a : α) :
    ((singleton a : PairingHeap α le).toListUnordered : Multiset α) = {a} :=
  rfl

instance : ToMultiset (PairingHeap α le) α where
  size x := x.size
  isEmpty x := x.isEmpty
  isEmpty_iff_size_eq_zero x :=
    match x with
    | ⟨.nil, _⟩ | ⟨.node _ _ _, _⟩ => by
      simp [isEmpty, size, PairingHeapImp.Heap.isEmpty, PairingHeapImp.Heap.size]
  toMultiset x := x.toListUnordered
  size_eq_card_toMultiset := by simp [size, toListUnordered]

instance : EmptyCollection (PairingHeap α le) where
  emptyCollection := empty

instance : LawfulEmptyCollection (PairingHeap α le) α where
  toMultiset_empty := rfl

@[simp]
lemma size_eq_zero_iff (x : PairingHeap α le) : x.size = 0 ↔ x = ∅ := by
  match x with
  | ⟨.nil, _⟩ => simpa [size] using by exact rfl
  | ⟨.node _ _ _, _⟩ => simpa [size] using nofun

instance : Insert α (PairingHeap α le) where
  insert := insert

instance : ToMultiset.LawfulInsert (PairingHeap α le) α where
  toMultiset_insert a x := by
    simp only [Insert.insert, insert, toMultiset]
    rw [coe_toListUnordered_merge, coe_toListUnordered_singleton]
    simp

instance : Mergeable (PairingHeap α le) α where
  merge := merge
  toMultiset_merge a x := by
    simp only [toMultiset, coe_toListUnordered_merge]

instance [Preorder α] [IsTotalPreorder α (· ≤ ·)] [DecidableRel (α := α) (· ≤ ·)] :
    MinHeap (PairingHeap α (· ≤ ·)) α where
  head? x := x.head?.rec ⊤ WithTop.some
  tail := tail
  deleteMin := deleteMin
  head?_eq_top x := by simpa [isEmpty_iff_size_eq_zero, Size.size] using by rintro rfl; rfl
  head?_mem_toMultiset_map x hx := by
    simp only [isEmpty_iff_size_eq_zero] at hx
    match x, hx with
    | ⟨.node a c .nil, _⟩, _ =>
      simpa using ⟨a, by
        simp [toMultiset, toListUnordered, PairingHeapImp.Heap.toListUnordered_node], rfl⟩
  head?_le x b hb := by
    match x with
    | ⟨.node a c .nil, hwf⟩ =>
      simp_rw [ToMultiset.mem_iff, toMultiset, toListUnordered,
        PairingHeapImp.Heap.toListUnordered_node, PairingHeapImp.Heap.toListUnordered_nil,
        List.append_nil, ← Multiset.cons_coe, Multiset.mem_cons] at hb
      obtain (rfl | hb) := hb; · rfl
      haveI : TotalBLE (α := α) (· ≤ ·) := ⟨by simp [total_of]⟩
      have := @PairingHeapImp.Heap.le_of_wf α (· ≤ ·) ⟨by simp⟩
        ⟨by simpa using fun _ _ _ ↦ le_trans⟩ _ a c .nil hwf
      simp? at this says
        simp only [PairingHeapImp.Heap.toListUnordered_node,
          PairingHeapImp.Heap.toListUnordered_nil, List.append_nil, List.mem_cons,
          decide_eq_true_eq, forall_eq_or_imp, le_refl, true_and] at this
      simpa [head?, PairingHeapImp.Heap.head?] using this b hb
  toMultiset_tail x h := by
    match x with
    | ⟨.nil, _⟩ => rfl
    | ⟨.node a c .nil, _⟩ =>
      simp? [Size.isEmpty, isEmpty, PairingHeapImp.Heap.isEmpty, tail, PairingHeapImp.Heap.tail,
          toMultiset, toListUnordered, PairingHeapImp.Heap.tail?, PairingHeapImp.Heap.deleteMin,
          - Multiset.coe_eq_coe, - Multiset.coe_erase] says
        simp only [toMultiset, toListUnordered, tail, PairingHeapImp.Heap.tail,
          PairingHeapImp.Heap.tail?, PairingHeapImp.Heap.deleteMin, Option.map_some',
          Option.getD_some, PairingHeapImp.Heap.coe_toListUnordered_combine, Size.isEmpty, isEmpty,
          PairingHeapImp.Heap.isEmpty, Bool.false_eq_true, ↓reduceDIte,
          PairingHeapImp.Heap.toListUnordered_node, PairingHeapImp.Heap.toListUnordered_nil,
          List.append_nil]
      change _ = Multiset.erase _ a
      simp
  deleteMin_def x := by match x with | ⟨.nil, _⟩ | ⟨.node _ _ _, _⟩ => rfl
