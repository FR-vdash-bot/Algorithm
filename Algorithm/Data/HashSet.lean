/-
Copyright (c) 2024 Yuyang Zhao. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Yuyang Zhao
-/
import Algorithm.Data.HashMap

variable {α : Type*} [BEq α] [Hashable α]

namespace Std.HashSet

instance : Size (HashSet α) where
  size := size
  isEmpty := isEmpty
  isEmpty_iff_size_eq_zero := Nat.beq_eq.to_iff

instance : LawfulEmptyCollection (HashSet α) α where
  not_mem_empty _ := not_mem_empty

variable [LawfulBEq α]

instance : ToFinset (HashSet α) α where
  toFinset c := ⟨c.toList, by simpa [List.Nodup] using HashMap.distinct_keys c.inner⟩
  mem_toFinset := by simp [toList]; rfl
  size_eq_card_toFinset := by simp [Size.size, toList, size]

instance : Erase (HashSet α) α where
  erase := erase

instance : LawfulErase (HashSet α) α := lawfulErase_iff_toFinset.mpr fun _ a ↦ by
  ext; simp [Erase.erase, mem_erase, eq_comm (a := a)]

instance : Bag (HashSet α) α where
  toFinset_insert a c {_} := by
    ext; simp [mem_insert, eq_comm (a := a)]

end Std.HashSet
