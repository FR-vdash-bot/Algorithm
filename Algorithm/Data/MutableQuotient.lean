/-
Copyright (c) 2024 Yuyang Zhao. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Yuyang Zhao
-/
import Mathlib.Data.Setoid.Basic
import Mutable.Mutable

/-!
# MutableQuotient

-/

universe u v w

namespace Quotient

def liftOnMkEq {α : Sort u} {β : Sort v} {s : Setoid α} (q : Quotient s)
    (f : ∀ a : α, ⟦a⟧ = q → β) (hf : ∀ a ha b hb, f a ha = f b hb) : β :=
  Equiv.subtypeQuotientEquivQuotientSubtype (fun a ↦ ⟦a⟧ = q) (· = q) (fun _ ↦ Iff.rfl)
    (fun _ _ ↦ Iff.rfl) ⟨q, rfl⟩ |>.lift
    (fun ⟨a, ha⟩ ↦ f a ha) (fun ⟨a, ha⟩ ⟨b, hb⟩ _ ↦ hf a ha b hb)

end Quotient

variable {M : Type u} {α : Type v} {β : Type w} {m : α → M}

def MutableQuotient (α : Type v) (m : α → M) :=
  Mutable (Quotient (Setoid.ker m))

namespace MutableQuotient

@[inline]
def mk (m : α → M) (a : α) : MutableQuotient α m := Mutable.mk ⟦a⟧

@[inline]
def get (x : MutableQuotient α m) (f : α → β) (hf : ∀ a₁ a₂, m a₁ = m a₂ → f a₁ = f a₂) : β :=
  (Mutable.get x).lift f fun _ _ h ↦ hf _ _ (by exact h)

@[simp]
lemma mk_get (m : α → M) (a : α) (f : α → β) (hf : ∀ a₁ a₂, m a₁ = m a₂ → f a₁ = f a₂) :
    (mk m a).get f hf = f a :=
  rfl

@[elab_as_elim, induction_eliminator]
lemma ind {motive : MutableQuotient α m → Prop} (h : ∀ (a : α), motive (mk m a)) (x) : motive x :=
  Quotient.ind (motive := fun x ↦ motive (Mutable.mk x)) h (Mutable.get x)

@[inline]
def getMkEq (x : MutableQuotient α m) (f : ∀ a : α, m a = x.get m (fun _ _ ↦ id) → β)
    (hf : ∀ a₁ ha₁ a₂ ha₂, f a₁ ha₁ = f a₂ ha₂) : β :=
  (Mutable.get x).liftOnMkEq (fun a ha ↦ f a (congr_arg (Quotient.lift m _) ha))
    (fun _ _ _ _ ↦ hf _ _ _ _)

@[inline]
def map (x : MutableQuotient α m) (f : α → α) (hf : ∀ a₁ a₂, m a₁ = m a₂ → m (f a₁) = m (f a₂)) :
    MutableQuotient α m :=
  get x (fun a ↦ mk m (f a)) (Mutable.ext <| Quotient.sound <| hf · · ·)

-- def getModify (x : MutableQuotient α m) (f : α → β) (hf : ∀ a₁ a₂, m a₁ = m a₂ → f a₁ = f a₂)
--     (r : α → α) (hr : ∀ a, m (r a) = m a) : β :=
--   (Mutable.getModify x (Quotient.map' r (fun a₁ a₂ h ↦ (hr a₁).trans h |>.trans (hr a₂).symm))
--     (Quotient.ind fun a ↦ Quotient.sound (hr a))).lift f fun _ _ h ↦ hf _ _ (by exact h)

@[inline]
def getModify (x : MutableQuotient α m) (fr : α → β × α)
    (hf : ∀ a₁ a₂, m a₁ = m a₂ → (fr a₁).fst = (fr a₂).fst) (hr : ∀ a, m (fr a).snd = m a) : β :=
  Mutable.getModify₂ x (Quotient.lift (let ⟨x, y⟩ := fr ·; (x, ⟦y⟧)) (fun a₁ a₂ h ↦
      Prod.ext (hf a₁ a₂ h) <| Quotient.sound <| (hr a₁).trans <| h.trans (hr a₂).symm))
    (Quotient.ind fun a ↦ Quotient.sound (hr a))

@[simp]
lemma mk_getModify (m : α → M) (a : α) (fr : α → β × α)
    (hf : ∀ a₁ a₂, m a₁ = m a₂ → (fr a₁).fst = (fr a₂).fst) (hr : ∀ a, m (fr a).snd = m a) :
    (mk m a).getModify fr hf hr = (fr a).fst :=
  rfl

end MutableQuotient
