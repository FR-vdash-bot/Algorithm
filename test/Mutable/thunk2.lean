import Algorithm.Data.Mutable

universe u

def Thunk' (α : Type u) : Type u := Mutable (Unit → α)

variable {α : Type u}

def Thunk'.mk (fn : Unit → α) : Thunk' α := Mutable.mk fn

protected def Thunk'.pure (a : α) : Thunk' α :=
  .mk fun _ ↦ a

protected def Thunk'.get (x : Thunk' α) : α :=
  Mutable.getWith₂ x (fun f ↦ f ()) (fun a _ ↦ a) (fun _ ↦ rfl)

/-! lean4/tests/compiler/thunk.lean -/

def compute (v : Nat) : Thunk' Nat :=
  .mk (fun _ => let xs := List.replicate 100000 v; xs.foldl Nat.add 0)

@[noinline]
def test (t : Thunk' Nat) (n : Nat) : Nat :=
  n.repeat (fun r => t.get + r) 0

def main : IO UInt32 :=
  IO.println (toString (test (compute 1) 100000)) *>
  pure 0

#eval main
