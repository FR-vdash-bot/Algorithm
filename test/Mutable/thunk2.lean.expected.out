10000000000
0
